File: src/minions/memory/schema.py
"""
Memory Schema - Unified schema for multi-agent memory system.

Defines the standard format for memories across all agents.
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any


class MemoryScope(str, Enum):
    """Memory visibility scope."""

    SESSION = "session"  # Current session only (temporary)
    PROJECT = "project"  # Project-specific, persistent
    USER = "user"  # User-wide, persistent
    AGENT = "agent"  # Specific agent only
    PUBLIC = "public"  # Shared across all agents


class MemoryType(str, Enum):
    """Type of memory event."""

    OBSERVATION = "observation"  # Factual observation
    DECISION = "decision"  # Design/implementation decision
    PLAN = "plan"  # Future plan or intent
    ARTIFACT = "artifact"  # Code, file, or output reference
    PREFERENCE = "preference"  # User preference
    WORKFLOW = "workflow"  # Workflow pattern
    ERROR = "error"  # Error pattern and solution
    RESEARCH = "research"  # Research finding


class AgentType(str, Enum):
    """Agent identifiers."""

    CLAUDE = "claude"
    CODEX = "codex"
    GEMINI = "gemini"
    COPILOT = "copilot"
    SYSTEM = "system"  # For system-generated memories


@dataclass
class MemoryEvent:
    """
    Unified memory event schema.

    All memories across all agents conform to this schema.
    """

    # Required fields
    content: str
    memory_type: MemoryType
    scope: MemoryScope
    source_agent: AgentType

    # Auto-generated fields
    id: str = field(default_factory=lambda: datetime.now().strftime("%Y%m%d%H%M%S%f"))
    created_at: str = field(default_factory=lambda: datetime.now().isoformat())

    # Optional fields
    context: str = ""
    confidence: float = 1.0  # 0.0 to 1.0
    ttl_days: int | None = None  # Time to live, None = permanent
    tags: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)

    # Embedding (populated by mem0)
    embedding: list[float] | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "id": self.id,
            "content": self.content,
            "memory_type": self.memory_type.value,
            "scope": self.scope.value,
            "source_agent": self.source_agent.value,
            "context": self.context,
            "confidence": self.confidence,
            "ttl_days": self.ttl_days,
            "tags": self.tags,
            "metadata": self.metadata,
            "created_at": self.created_at,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "MemoryEvent":
        """Create from dictionary."""
        return cls(
            id=data.get("id", ""),
            content=data["content"],
            memory_type=MemoryType(data["memory_type"]),
            scope=MemoryScope(data["scope"]),
            source_agent=AgentType(data["source_agent"]),
            context=data.get("context", ""),
            confidence=data.get("confidence", 1.0),
            ttl_days=data.get("ttl_days"),
            tags=data.get("tags", []),
            metadata=data.get("metadata", {}),
            created_at=data.get("created_at", ""),
        )


# Sensitive patterns to redact
SENSITIVE_PATTERNS = [
    r"sk-[a-zA-Z0-9_-]{20,}",  # OpenAI API key (various formats)
    r"sk-proj-[a-zA-Z0-9_-]+",  # OpenAI project API key
    r"sk-ant-[a-zA-Z0-9\-]{20,}",  # Anthropic API key
    r"AIza[a-zA-Z0-9_-]{35}",  # Google API key
    r"ghp_[a-zA-Z0-9]{36}",  # GitHub token
    r"gho_[a-zA-Z0-9]{36}",  # GitHub OAuth token
    r"password\s*[:=]\s*\S+",  # Password patterns
    r"secret\s*[:=]\s*\S+",  # Secret patterns
]

File: resources/hooks-rs/crates/hook-memory/src/schema.rs
//! Memory schema - unified format for multi-agent memory system.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Memory visibility scope.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MemoryScope {
    /// Current session only
    Session,
    /// User-wide, persistent
    User,
    /// Specific agent only
    Agent,
    /// Shared across all agents
    Public,
}

/// Type of memory event.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MemoryType {
    /// Factual observation
    Observation,
    /// Design/implementation decision
    Decision,
    /// Future plan or intent
    Plan,
    /// Code, file, or output reference
    Artifact,
    /// User preference
    Preference,
    /// Workflow pattern
    Workflow,
    /// Error pattern and solution
    Error,
    /// Research finding
    Research,
}

/// Agent identifiers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AgentType {
    Claude,
    Codex,
    Gemini,
    Copilot,
    System,
}

/// Unified memory event schema.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryEvent {
    /// Unique identifier
    pub id: String,

    /// Memory content
    pub content: String,

    /// Type of memory
    pub memory_type: MemoryType,

    /// Visibility scope
    pub scope: MemoryScope,

    /// Source agent
    pub source_agent: AgentType,

    /// Additional context
    #[serde(default)]
    pub context: String,

    /// Confidence score (0.0 to 1.0)
    #[serde(default = "default_confidence")]
    pub confidence: f64,

    /// Time to live in days (None = permanent)
    #[serde(default)]
    pub ttl_days: Option<u32>,

    /// Tags for categorization
    #[serde(default)]
    pub tags: Vec<String>,

    /// Additional metadata
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,

    /// Creation timestamp (ISO 8601)
    pub created_at: String,
}

fn default_confidence() -> f64 {
    1.0
}

impl MemoryEvent {
    /// Create a new memory event.
    pub fn new(
        content: impl Into<String>,
        memory_type: MemoryType,
        scope: MemoryScope,
        source_agent: AgentType,
    ) -> Self {
        let now = chrono_now();
        Self {
            id: generate_id(),
            content: content.into(),
            memory_type,
            scope,
            source_agent,
            context: String::new(),
            confidence: 1.0,
            ttl_days: None,
            tags: Vec::new(),
            metadata: HashMap::new(),
            created_at: now,
        }
    }

    /// Set context.
    pub fn with_context(mut self, context: impl Into<String>) -> Self {
        self.context = context.into();
        self
    }

    /// Set confidence.
    pub fn with_confidence(mut self, confidence: f64) -> Self {
        self.confidence = confidence.clamp(0.0, 1.0);
        self
    }

    /// Set TTL.
    pub fn with_ttl(mut self, days: u32) -> Self {
        self.ttl_days = Some(days);
        self
    }

    /// Add tag.
    pub fn with_tag(mut self, tag: impl Into<String>) -> Self {
        self.tags.push(tag.into());
        self
    }
}

/// Generate a unique ID based on timestamp.
fn generate_id() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let duration = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    format!("{}{:06}", duration.as_secs(), duration.subsec_micros())
}

/// Get current time in ISO 8601 format.
fn chrono_now() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let duration = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    let secs = duration.as_secs();
    // Simple ISO 8601 format (without timezone)
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}",
        1970 + secs / 31536000,
        (secs % 31536000) / 2592000 + 1,
        (secs % 2592000) / 86400 + 1,
        (secs % 86400) / 3600,
        (secs % 3600) / 60,
        secs % 60
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_memory_event_creation() {
        let event = MemoryEvent::new(
            "Test memory",
            MemoryType::Observation,
            MemoryScope::User,
            AgentType::Claude,
        );
        assert_eq!(event.content, "Test memory");
        assert_eq!(event.memory_type, MemoryType::Observation);
        assert_eq!(event.confidence, 1.0);
    }

    #[test]
    fn test_memory_event_builder() {
        let event = MemoryEvent::new(
            "Preference",
            MemoryType::Preference,
            MemoryScope::User,
            AgentType::System,
        )
        .with_context("User said")
        .with_confidence(0.9)
        .with_tag("pr")
        .with_tag("japanese");

        assert_eq!(event.context, "User said");
        assert_eq!(event.confidence, 0.9);
        assert_eq!(event.tags, vec!["pr", "japanese"]);
    }

    #[test]
    fn test_serialization() {
        let event = MemoryEvent::new(
            "Test",
            MemoryType::Decision,
            MemoryScope::Session,
            AgentType::Codex,
        );
        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains("\"memory_type\":\"decision\""));
        assert!(json.contains("\"scope\":\"session\""));
    }
}
